### ISSUE TITLE
Error when creating arrays with more than 1 dimension

### ISSUE THREAD

**Issue by gbaydin** [arrayfire/arrayfire-dotnet#3]
Hi!

I'm really happy to have found this ArrayFire wrapper for .NET. Thank you for providing this code.

I have a difficulty converting an existing array into an `ArrayFire.Array`, when it's not 1 dimensional.

When I do this, in F#:

```fsharp
let a = [|1.; 2.; 3.; 4.|]
let aa = ArrayFire.Data.CreateArray(a)
```

everything is fine and I can successfully use this array in subsequent operations. But when I try to create an `ArrayFire.Array` with two dimensions,

```fsharp
let b = array2D [[1.; 2.]; [3.; 4.]]
let bb = ArrayFire.Data.CreateArray(b)
```

I get an "invalid input size" exception

`ArrayFire.ArrayFireException: Invalid input size at ArrayFire.Data.CreateArray(Double[,] data)`

It also fails with 3D arrays such as this:

```fsharp
let c = Array3D.zeroCreate<float> 2 2 2
let cc = ArrayFire.Data.CreateArray(c)
```

**Comment by shehzan10** [comment]
I should mention that this wrapper isn't complete and is a work in progress.
On the question you raised, you can look at the fsharp example that we have https://github.com/arrayfire/arrayfire-dotnet/blob/devel/Examples/HelloWorld/FSharp/Program.fs for syntax.

**Comment by gbaydin** [comment]
Hi @shehzan10, thank you for the quick answer.

This is about a very essential functionality. The wrapper has code in place for handling arrays up to 4D. It just doesn't seem to work. Or I'm not using it in the intended way.

I understand that this wrapper is in a preliminary stage. I'm very interested in using it, and I can also try to contribute a fix for this, if you point me in the right direction.

Creating random 2D arrays, as in the example you mentioned

```fsharp
let arr1 = Data.RandNormal<double>(3, 3)
```
works fine.

I need a way of converting an existing 2D array into an `ArrayFire.Array`, for making use of the library.

**Comment by gbaydin** [comment]
I found the bug.

The error is in 

https://github.com/arrayfire/arrayfire-dotnet/blob/devel/Wrapper/Data.cs

where the dimension information of calls to `AFArray.af_create_array` are passed incorrectly.

For example, in line 86,

the code
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static Array CreateArray(float[,] data) { IntPtr ptr; Internal.VERIFY(AFArray.af_create_array(out ptr, data, (uint)data.Rank, new long[] { data.Length }, af_dtype.f32)); return new Array(ptr); }
```

should be corrected as
```csharp
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static Array CreateArray(float[,] data) { IntPtr ptr; Internal.VERIFY(AFArray.af_create_array(out ptr, data, (uint)data.Rank, new long[] { data.GetLength(0), data.GetLength(1) }, af_dtype.f32)); return new Array(ptr); }
```

The problem is with the `long[] dim_dims` parameter.

I think this code is generated by the AutoGenTool included in the repository. The rules for larger than 1D arrays should be fixed there.

**Comment by gbaydin** [comment]
I submitted a pull request that fixes the problem with `Data.CreateArray` and also with `Data.GetData`. My code runs fine with these changes.

**Comment by royalstream** [comment]
Hi guys, sorry for the late reply, I was out on vacation. 

Thanks for finding this and patching it. I added *host<->device* copying for 2/3/4-dimensional as an afterthought and I never got to test it.

There's no need to update the *AutoGenTool* because this tool regenerates the `AFArray.af_*` classes (which didn't change) and the changes inside *Data.cs* get picked up automatically.

**Comment by altaybrusan** [comment]
Hi,
I could not find a way to select a sub section of an array object and fill it by another array. Is it feasible or not yet?

**Comment by royalstream** [comment]
I did create something similar to the C++ proxy object to support that behavior but on a local experimental branch. I'm going to finish it and create a pull-request by the end of tomorrow (Monday). Hope that helps.

**Comment by altaybrusan** [comment]
Thanks

**Comment by royalstream** [comment]
@altaybrusan I just created the Pull Request I promised.

@shehzan10 please help me closing this issue, it was about something that got fixed a while ago.

**Comment by shehzan10** [comment]
Fixed by #7
